### PR TITLE
Add BoxSelectTool callback initial commit

### PR DESCRIPTION
**Pull request by canavandl**
ref issue: #2465 

Add callback on BoxSelectTool pan end event.

### CODE REVIEW

**Comment by canavandl** [comment]
So I changed the User Guide example from the drawing_rects.png one to the stock range_selection.png. I've now realized that it's possible and maybe better to do the stock timeseries one with simply a callback on the selection (and not on the box select tool), so I don't know if I should go back to the rects example. Thoughts?

**Comment by birdsarah** [comment]
Personally I really like the box example. I felt like Mondrian clicking away. (next we can add an example where you can pick your color too). I think the we have enough examples with stock market data.

**Comment by canavandl** [comment]
I also like the box example, so I'll roll back the branch. Thanks for the feedback.

File: sphinx/source/docs/user_guide/interaction.rst [hunk 1]
```diff
@@ -204,6 +204,68 @@ changes the source of a plot when the slider is used.
...
+    output_file("boxselecttool_callback.html")
+
+    source = ColumnDataSource(data=dict(x=[], y=[], width=[], height=[]))
+
+    callback = Callback(args=dict(source=source), code="""
+        // get data source from Callback args
+        var data = source.get('data');
+
+        /// get BoxSelectTool dimensions from cb_data parameter of Callback
+        var geometry = cb_data['geometry']
```
Comment (birdsarah) [inline sphinx/source/docs/user_guide/interaction.rst]: Would be nice, if all this js had ; at the end of the line.

File: sphinx/source/docs/user_guide/interaction.rst [hunk 2]
```diff
@@ -204,6 +204,68 @@ changes the source of a plot when the slider is used.
...
+have access to the tool's attributes. Below, a callback on the BoxSelectTool
+uses the selection box dimensions (accessed in the geometry field of the
+cb_data object that is injected into the Callback code attribute), in order to
+add a Rect glyph to the plot with identical dimensions.
+
+.. bokeh-plot::
+    :source-position: above
+
+    from bokeh.models import Callback, ColumnDataSource, BoxSelectTool, Range1d, Rect
+    from bokeh.plotting import hplot, figure, output_file, show
```
Comment (birdsarah) [inline sphinx/source/docs/user_guide/interaction.rst]: hplot is unused.

File: bokehjs/src/coffee/tool/gestures/box_select_tool.coffee [hunk 3]
```diff
@@ -62,10 +62,34 @@ class BoxSelectToolView extends SelectTool.View
...
 
+    if @mget('callback')?
+      @_emit_callback(geometry)
+
     @_save_geometry(geometry, final, append)
 
     return null
 
+  _emit_callback: (geometry) ->
+    r = @mget('renderers')[0]
```
Comment (birdsarah) [inline bokehjs/src/coffee/tool/gestures/box_select_tool.coffee]: This is just a question for my own learning - why do you take the first renderer here?

**Comment by birdsarah** [comment]
I've looked at the code, run the docs, and run and played with the example locally in py34. It looks good to me.

**Comment by canavandl** [comment]
I'll removed the unused hplot import and improve the JS style before merging

File: bokeh/models/tools.py [hunk 4]
```diff
@@ -192,8 +192,8 @@ class CrosshairTool(Tool):
 
     dimensions = List(Enum(Dimension), default=["width", "height"], help="""
     Which dimensions the crosshair tool is to track. By default, both a
-    vertical and horizontal line will be dran. If only "width" is supplied, 
-    only a horizontal line will be drawn. If only "height" is supplied, 
+    vertical and horizontal line will be dran. If only "width" is supplied,
```
Comment (damianavila) [inline bokeh/models/tools.py]: `drawn`

**Comment by damianavila** [comment]
I am seeing warnings in the docs building:

```
/media/data/devel/continuum/bokeh_project/bokeh/bokeh/properties.py:docstring of bokeh.properties.value:23: ERROR: Unexpected indentation.
/media/data/devel/continuum/bokeh_project/bokeh/sphinx/source/docs/user_guide/styling.rst:81: WARNING: Duplicate explicit target name: "userguide_styling_colors".
```

Someone else is seeing them?

**Comment by damianavila** [comment]
Also this is not merging cleanly...

**Comment by canavandl** [comment]
I fixed the merge conflict and the duplicate target name warn but couldn't figure out the indentation one. When the build passes, this can be merged.

**Comment by damianavila** [comment]
>I fixed the merge conflict and the duplicate target name warn but couldn't figure out the indentation one. 

We can see to fix it later in follow up docs PR after the freeze...

**Pull request merged**